extensions/librdkafka/ConsumeKafka.cpp

```diff
@@ -0,0 +1,553 @@
+  }
+}
+
+void ConsumeKafka::extend_config_from_dynamic_properties(const core::ProcessContext* context) {
+  using utils::setKafkaConfigurationField;
+
+  const std::vector<std::string> dynamic_prop_keys = context->getDynamicPropertyKeys();
+  if (dynamic_prop_keys.empty()) {
+    return;
+  }
+  logger_->log_info("Loading %d extra kafka configuration fields from ConsumeKafka dynamic properties:", dynamic_prop_keys.size());
+  for (const std::string& key : dynamic_prop_keys) {
+    std::string value;
+    gsl_Expects(context->getDynamicProperty(key, value));
+    logger_->log_info("%s: %s", key.c_str(), value.c_str());
+    setKafkaConfigurationField(conf_.get(), key, value);
+  }
+}
+
+void ConsumeKafka::configure_new_connection(const core::ProcessContext* context) {
+  using utils::setKafkaConfigurationField;
+
+  conf_ = { rd_kafka_conf_new(), utils::rd_kafka_conf_deleter() };
+  if (conf_ == nullptr) {
+    throw Exception(PROCESS_SCHEDULE_EXCEPTION, "Failed to create rd_kafka_conf_t object");
+  }
+
+  // Set rebalance callback for use with coordinated consumer group balancing
+  // Rebalance handlers are needed for the initial configuration of the consumer
+  // If they are not set, offset reset is ignored and polling produces messages
+  // Registering a rebalance_cb turns off librdkafka's automatic partition assignment/revocation and instead delegates that responsibility to the application's rebalance_cb.
+  rd_kafka_conf_set_rebalance_cb(conf_.get(), rebalance_cb);
+
+  // Uncomment this for librdkafka debug logs:
+  // setKafkaConfigurationField(conf_.get(), "debug", "all");
+
+  setKafkaConfigurationField(conf_.get(), "bootstrap.servers", kafka_brokers_);
+  setKafkaConfigurationField(conf_.get(), "auto.offset.reset", "latest");
+  setKafkaConfigurationField(conf_.get(), "enable.auto.commit", "false");
+  setKafkaConfigurationField(conf_.get(), "enable.auto.offset.store", "false");
+  setKafkaConfigurationField(conf_.get(), "isolation.level", honor_transactions_ ? "read_committed" : "read_uncommitted");
+  setKafkaConfigurationField(conf_.get(), "group.id", group_id_);
+  setKafkaConfigurationField(conf_.get(), "session.timeout.ms", std::to_string(session_timeout_milliseconds_.count()));
+  setKafkaConfigurationField(conf_.get(), "max.poll.interval.ms", "600000");  // Twice the default, arbitrarily chosen
+
+  // This is a librdkafka option, but the communication timeout is also specified in each of the
+  // relevant API calls. Could be redundant, but it probably does not hurt to set this
+  setKafkaConfigurationField(conf_.get(), "metadata.request.timeout.ms", std::to_string(METADATA_COMMUNICATIONS_TIMEOUT_MS));
+
+  extend_config_from_dynamic_properties(context);
+
+  std::array<char, 512U> errstr{};
+  consumer_ = { rd_kafka_new(RD_KAFKA_CONSUMER, conf_.release(), errstr.data(), errstr.size()), utils::rd_kafka_consumer_deleter() };
+  if (consumer_ == nullptr) {
+    const std::string error_msg { errstr.begin(), errstr.end() };
```

Comment:
This will create a string of length 512 with embedded nulls.  If we are sure Kafka returns a null-terminated error message, we can do `const std::string error_msg{errstr.data()};`

Reply:
Actually, I am not sure it is a safe assumption.

Reply:
I'm pretty sure it's null-terminated, because C strings are null-terminated and because the API has no way of returning the number of characters written to the error buffer.

Reply:
Updated along with the other occurences.